src/librustc_resolve/lib.rs

```diff
@@ -84,6 +84,13 @@ mod check_unused;
 mod build_reduced_graph;
 mod resolve_imports;
 
+pub struct ResolvePath<'a>
+{
+    ident : &'a [Ident],
+    source : Option<NodeId>,    // None if this path is speculative
+    speculative : bool,
```

Comment:
There currently isn't a case where this is true and source is None, let's remove this field. Instead, document the source as 
"NodeId of the path that we are attempting to resolve. When None, this path is being speculatively resolved and we should not emit errors or lints about it."